futures-util/src/future/shared.rs

```diff
@@ -229,26 +246,32 @@ impl<Fut: Future> Future for Shared<Fut> {
                         _ => unreachable!(),
                     }
                 }
-                Poll::Ready(output) => {
-                    let output = Arc::new(output);
-                    unsafe {
-                        *this.inner.future_or_output.get() =
-                            FutureOrOutput::Output(output.clone());
-                    }
+                Poll::Ready(output) => break output,
+            }
+        };
 
-                    // Complete the future
-                    let mut lock = this.inner.notifier.wakers.lock().unwrap();
-                    this.inner.notifier.state.store(COMPLETE, SeqCst);
-                    let wakers = &mut lock.take().unwrap();
-                    for (_key, opt_waker) in wakers {
-                        if let Some(waker) = opt_waker.take() {
-                            waker.wake();
-                        }
-                    }
-                    return Poll::Ready(output);
-                }
+        if let Some(_) = Arc::get_mut(&mut this.inner) {
```

Comment:
Had to move all this code out here since `this.inner` was borrowed by `Reset`. Does the code after this check also need to be covered by `Reset`?

Reply:
Yes-- can you move it all into a block containing the `Reset`?